src/audio/drc/drc_generic.c

```diff
@@ -450,23 +490,227 @@ static void drc_s16_default(const struct comp_dev *dev,
+					pd_read++;
+				}
+				idx += nch;
+			}
+		}
+
+		state->pre_delay_write_index += frames;
+		state->pre_delay_write_index &= DRC_MAX_PRE_DELAY_FRAMES_MASK;
+		state->pre_delay_read_index += frames;
+		state->pre_delay_read_index &= DRC_MAX_PRE_DELAY_FRAMES_MASK;
+		return;
+	}
+
+	if (!state->processed) {
+		drc_update_envelope(state, p);
+		drc_compress_output(state, p, 4, nch);
+		state->processed = 1;
+	}
+
+	offset = state->pre_delay_write_index & DRC_DIVISION_FRAMES_MASK;
+	while (i < frames) {
+		/* Copy fragment data from source to pre-delay buffers, and copy the output fragment
+		 * to sink */
+		fragment = MIN(DRC_DIVISION_FRAMES - offset, frames - i);
+		pd_write_index = state->pre_delay_write_index;
+		pd_read_index = state->pre_delay_read_index;
+		for (ch = 0; ch < nch; ++ch) {
+			pd_write = (int32_t *)state->pre_delay_buffers[ch] + pd_write_index;
+			pd_read = (int32_t *)state->pre_delay_buffers[ch] + pd_read_index;
+			idx = i * nch + ch;
+			for (f = 0; f < fragment; ++f) {
+				x = audio_stream_read_frag_s32(source, idx);
+				y = audio_stream_read_frag_s32(sink, idx);
+				*pd_write = *x << 8; /* write pre_delay_buffer as s32 formnat */
+				*y = sat_int24(Q_SHIFT_RND(*pd_read, 31, 23)); /* read pre_delay_buffer as s32 format */
+				pd_write++;
+				pd_read++;
+				idx += nch;
+			}
+		}
+		state->pre_delay_write_index = (pd_write_index + fragment) & DRC_MAX_PRE_DELAY_FRAMES_MASK;
+		state->pre_delay_read_index = (pd_read_index + fragment) & DRC_MAX_PRE_DELAY_FRAMES_MASK;
+
+		i += fragment;
+		offset = (offset + fragment) & DRC_DIVISION_FRAMES_MASK;
+
+		/* Process the input division (32 frames). */
+		if (offset == 0)
+			drc_process_one_division(state, p, 4, nch);
+	}
+}
+#endif /* CONFIG_FORMAT_S24LE */
+
+#if CONFIG_FORMAT_S32LE
+static void drc_s32_default(const struct comp_dev *dev,
+			    const struct audio_stream *source,
+			    struct audio_stream *sink,
+			    uint32_t frames)
+{
+	int32_t *x;
+	int32_t *y;
+	int32_t *pd_write;
+	int32_t *pd_read;
+	int offset;
+	int i = 0;
+	int ch;
+	int idx;
+	int f;
+	int fragment;
+	int pd_write_index;
+	int pd_read_index;
+	int nch = source->channels;
+
+	struct drc_comp_data *cd = comp_get_drvdata(dev);
+	struct drc_state *state = &cd->state;
+	const struct sof_drc_params *p = &cd->config->params; /* Read-only */
+
+	if (!p->enabled) {
+		/* Delay the input sample only and don't do other processing. This is used when the
+		 * DRC is disabled. We want to do this to match the processing delay of other bands
+		 * in multi-band DRC kernel case. */
+		for (ch = 0; ch < nch; ++ch) {
+			pd_write_index = state->pre_delay_write_index;
+			pd_read_index = state->pre_delay_read_index;
+			pd_write = (int32_t *)state->pre_delay_buffers[ch] + pd_write_index;
+			pd_read = (int32_t *)state->pre_delay_buffers[ch] + pd_read_index;
+			idx = ch;
+			for (i = 0; i < frames; ++i) {
+				x = audio_stream_read_frag_s32(source, idx);
+				y = audio_stream_read_frag_s32(sink, idx);
```

Comment:
Should this be audio_stream_write_frag_s32() and the same for previous function for s24 format?